### PR TITLE
Add sets package

### PR DESCRIPTION
**Pull request by dnezam**
**Roadmap**
- [x] Port over desired definitions from Dafny
- [x] Port over desired definitions from Verus
- [x] Port over desired definitions from Why3
- [x] Discuss whether additional definitions must be added to account for disabling the axiomization in user packages
- [x] Decide whether to use standard libraries of proof assistants like Coq and Lean
- [x] Restructure package to make it consistent

### CODE REVIEW

**Review by jcp19 — CHANGES_REQUESTED** [review]
Some comments for now, I will continue later

File: sets/sets.gobra
Comment (dnezam) [reply sets/sets.gobra]: If I add the // ##(-I ../) annotation it works, but I couldn't figure out how to make it work without the annotation. Passing -I "../" or -I ../ didn't work. Also, I am not sure what the syntax is to add a path to includePaths (currently ".") in the gobra.yml file.

File: sets/sets.gobra
Comment (dnezam) [reply sets/sets.gobra]: Do you mean ChooseFromSingleton or Choose? For now, I have removed ChooseFromSingleton and made Choose private.

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:10]:
```suggestion
  This file is inspired by the standard libraries and axiomatisations of the following verifiers:
  - dafny-lang/libraries: https://github.com/dafny-lang/libraries/blob/master/src/Collections/Sets/Sets.dfy
  - verus-lang/verus: https://github.com/verus-lang/verus/blob/main/source/pervasive/set_lib.rs
  - why3: https://why3.lri.fr/stdlib/set.html
  - viperproject/silicon: https://github.com/viperproject/silicon/blob/master/src/main/resources/dafny_axioms/sets.vpr
```

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:21]: I will leave it up to you to decide, and I will let you know if I think some description is missing

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:33]:
```suggestion
// Returns the empty set.
```

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:25]: I would drop these headers across the file

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:270]: That would be ok. The alternative would be to have a separate lemma for that

File: sets/sets.gobra
Comment (jcp19) [reply sets/sets.gobra:327]: (Make sure that the spacing is consistent)

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:362]: I would merge them

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:385]: Same here

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:422]: Let's talk about it in the next meeting

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:515]: let's discuss it in the next meeting

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:520]: what do these functions do?

File: sets/sets.gobra
Comment (jcp19) [inline sets/sets.gobra:529]: In general, grouping postconditions like this seems desirable, given that we would not need to call additional lemmas to prove "obvious/natural" properties of the returned values.

**Review by jcp19 — APPROVED** [review]
Ready to merge after the files are cleaned up and my comments are addressed

File: sets/sets.gobra
Comment (dnezam) [reply sets/sets.gobra:520]: min_elt returns the smallest element in a non-empty set, while max_elt returns the largest element in a non-empty set.

**Pull request merged**
